### PR TITLE
Implement sampling

### PR DESCRIPTION
**Pull request by mrober**

### CODE REVIEW

File: firebase-sessions/src/main/kotlin/com/google/firebase/sessions/FirebaseSessions.kt [hunk 1]
```diff
@@ -62,9 +62,19 @@ internal constructor(
...
+    if (!sessionDetails.collectEvents) {
+      Log.d(TAG, "Session Start event sampled.")
+      return
     }
+
+    sessionCoordinator.attemptLoggingSessionEvent(sessionEvent)
+  }
+
+  /** Calculate whether we should sample events using [sessionSettings] data. */
+  private fun shouldCollectEvents(): Boolean {
```
Comment (visumickey) [inline firebase-sessions/src/main/kotlin/com/google/firebase/sessions/FirebaseSessions.kt:74]: Just want to make clear that we are not sampling events for every session. Sampling should be implemented by a cold start. PTAL at the design doc for sampling in the AQS shared folder.

Comment (mrober) [reply firebase-sessions/src/main/kotlin/com/google/firebase/sessions/FirebaseSessions.kt:74]: Done

File: firebase-sessions/src/main/kotlin/com/google/firebase/sessions/FirebaseSessions.kt [hunk 2]
```diff
@@ -35,11 +35,11 @@ internal constructor(
   backgroundDispatcher: CoroutineDispatcher,
   transportFactoryProvider: Provider<TransportFactory>,
 ) {
-  private val sessionGenerator = SessionGenerator(collectEvents = true)
+  private val sessionSettings = SessionsSettings(firebaseApp.applicationContext)
```
Comment (visumickey) [inline firebase-sessions/src/main/kotlin/com/google/firebase/sessions/FirebaseSessions.kt:38]: This is good. But we need to find out a way to wait for some seconds before initiating a network config fetch in the settings layer to fetch configs.

Comment (mrober) [reply firebase-sessions/src/main/kotlin/com/google/firebase/sessions/FirebaseSessions.kt:38]: Not enough of settings is implemented yet to do this now. I will keep this in mind though.

File: firebase-sessions/src/main/kotlin/com/google/firebase/sessions/FirebaseSessions.kt [hunk 3]
```diff
@@ -62,9 +62,19 @@ internal constructor(
     val sessionDetails = sessionGenerator.generateNewSession()
     val sessionEvent = SessionEvents.startSession(firebaseApp, sessionDetails, sessionSettings)
 
-    if (sessionDetails.collectEvents) {
-      sessionCoordinator.attemptLoggingSessionEvent(sessionEvent)
+    if (!sessionGenerator.collectEvents) {
```
[the remark this answers is not in the document]
Comment (mrober) [reply firebase-sessions/src/main/kotlin/com/google/firebase/sessions/FirebaseSessions.kt:65]: We still want to generate a session id to send to subscribers. I moved this around a bit so we don't build the whole session start event until the very end.

I will keep the settings in mind too. I imagine there will be some logic in `initiateSessionStart` to make sure we fetch settings.

File: firebase-sessions/src/main/kotlin/com/google/firebase/sessions/FirebaseSessions.kt [hunk 4]
```diff
@@ -62,9 +62,19 @@ internal constructor(
...
+      Log.d(TAG, "Session Start event sampled.")
```
[the remark this answers is not in the document]
Comment (mrober) [reply firebase-sessions/src/main/kotlin/com/google/firebase/sessions/FirebaseSessions.kt:67]: I added a TODO and a card to align on this.

**Review by samedson — APPROVED** [review]

File: firebase-sessions/src/main/kotlin/com/google/firebase/sessions/FirebaseSessions.kt (hunk 4 above)
[the remark this answers is not in the document]
Comment (samedson) [reply firebase-sessions/src/main/kotlin/com/google/firebase/sessions/FirebaseSessions.kt:67]: If you wanna do it right here, the message would be `Sessions SDK has sampled this session`, and debug is consistent

[the remark this answers is not in the document]
Comment (mrober) [reply firebase-sessions/src/main/kotlin/com/google/firebase/sessions/FirebaseSessions.kt]: Done

**Pull request merged**